lib/phoenix/socket/transport.ex

```diff
@@ -513,6 +511,18 @@ defmodule Phoenix.Socket.Transport do
     end
   end
 
+  defp check_cswsh(conn, session, csrf_token_key, :csrf_token) do
+    with csrf_token when is_binary(csrf_token) <- conn.params["_csrf_token"],
+         csrf_state when is_binary(csrf_state) <-
+           Plug.CSRFProtection.dump_state_from_session(session[csrf_token_key]) do
+      Plug.CSRFProtection.valid_state_and_csrf_token?(csrf_state, csrf_token)
+    end
+  end
+
+  defp check_cswsh(conn, _session, _csrf_token, :origin) do
+    conn.private[:phoenix_origin_checked]
```

Comment:
Now that I think about it, is this the wrong solution to the problem? This session is not valid and we should not assume it is. All this does is to set the `session` to `nil`. So what if instead of change Phoenix.LiveView instead so, if the LiveView explicitly allows such, it does not require the session to be valid (i.e. it accepts nil and converts it to an empty map)?

Reply:
I'm replying assuming you refer to make changes to LiveView as well. I mixed up goals in the initial explanation, so I'm going try to be clearer :sweat_smile: 

The goals are:
1. remove session data from Liveview's HTML to make them cacheable
2. remove the requirement of Liveview to have session data to enable visiting liveviews with or without cookie. This is a secondary goal and 1. is a prerequisite for it

To do that we would need:
a. to protect against websocket hijacking using `origin`, instead of session-based CSRF → this PR
b. make Liveview accept "empty" or `nil` session one way or another → not this PR

Now, because today we always check sessionfull liveviews with a CSRF token, there's always a session setup the:
```
            LiveView session was misconfigured or the user token is outdated.

            1) Ensure your session configuration in your endpoint is in a module attribute:

                @session_options [
                  ...
```
is displayed.

b. is not so important for now because there's always the possibility set a cookie before opening a liveview (via redirection).

> All this does is to set the session to nil. So what if instead of change Phoenix.LiveView so, if the LiveView explicitly allows such, it does not require the session to be valid (i.e. it accepts nil and converts it to an empty map)?

At first I changed the returned value here from `nil` to `%{}` when there's no session, but in that case LiveView will never show the above error message because there will always be an fake empty session. I'll make a separate PR with your proposed solution for `phoenix_live_view`. And I'll rename this PR as well.

Reply:
What I am saying is that I don’t think we need “a” (this PR). Browsers always send the origin and we always validate it if it is sent. Phoenix is already safe from CSWHJ. The session code is only validating if the session is valid, nothing more.

Reply:
> The session code is only validating if the session is valid, nothing more.

Might be wrong but I think the session code in socket/transport performs CSRF check to prevent websocket hjiacking (CSWSH) specifically. There's no need to check CSRF for GET request for instance, right?

> Browsers always send the origin and we always validate it if it is sent.

If we assume that (which is true in 99.99% of cases I guess), then we can:
- let `origin` check as it is today
- remove the CSRF token check in the session code (if my assumption about the initial goal of its implementation is true)

What do you think?

Reply:
> Might be wrong but I think the session code in socket/transport performs CSRF check to prevent websocket hjiacking (CSWSH) specifically. There's no need to check CSRF for GET request for instance, right?

The session+csrf code checks if the session is valid/invalid. You can use the fact if the session is valid or invalid to protect against CSWSH by failing to accept connections with a nil session. My point is that the CSRF code here does not cancel and abort the request in any way, it leaves this decision to a future layer.

Reply:
Yes, but my point is that you can't fetch the session without CSRF token, and checking against CSWSH with a CSRF token makes the liveview non-cacheable (in a shared = public cache), as the token is linked to the user session.

This is why I'm looking to implement a mechanism that:
- protects against CSWSH → using the `origin` header instead
- allows fetching the session → even in the absence of a CSRF token
- doesn't set private data in Liveview's initial render HTML → no use of a CSRF token embedded in the HTML

> My point is that the CSRF code here does not cancel and abort the request in any way

It does prevent the Liveview from mounting if user A request first the page which is cached and then user B fetches the cached page: CSRF fails to validate.

> The session+csrf code checks if the session is valid/invalid.

I'd say the plug session store checks the validity of the cookie (signature, expiration) and I believe CSRF in this case specifically protects against CSWSH. [
Plug.CSRFProtection ](https://hexdocs.pm/plug/Plug.CSRFProtection.html) says:

> GET requests are not protected, as they should not have any side-effect or change your application state. JavaScript requests are an exception: by using a script tag, external websites can embed server-side generated JavaScript, which can leak information. For this reason, this plug also forbids any GET JavaScript request that is not XHR (or AJAX)

which makes me think that since Liveview's request is a `GET` request.

Reply:
Ok, so the part we are disagreeing is this:

> allows fetching the session → even in the absence of a CSRF token

Why would want to fetch the session in a LiveView that was publicly cached?

This seems like a very dangerous place to be because, if you are relying on a session existing in the first place, then the page you are caching (which was most likely generated by the same LiveView) is using parts of the session.

So my point is: if you want to cache it, then the safest thing for us to do is to make sure you can't access the session at all. Not in the dead render, nor the live render.

> It does prevent the Liveview from mounting if user A request first the page which is cached and then user B fetches the cached page: CSRF fails to validate.

Yes, but this is LiveView logic, no Phoenix logic. :)

Reply:
> Why would want to fetch the session in a LiveView that was publicly cached?

That's the *publicly-cacheable private LiveViews* use-case: in many cases only a few few elements are "private" (session-based). For example on my online shop's landing page the only session-based elements are: my name, the number of liked products and the number of products in the basket. The idea is to:
- cache the initial Liveview render - the page only contains public data
- let Liveview update the page with public & private elements through websocket

> This seems like a very dangerous place to be

Definitely! So many ways to shoot yourself in the foot. However this is a pattern that's already being used:
- varnish + ESI (Edge Side Includes) that do just that
- some commercial products that seem to use either that or ws

Since I published [Introducing `http_cache`, a BEAM-native standard-compliant HTTP caching library](http://svground.fr/blog/posts/introducing-http-cache/), I received a few messages from people wanting to do exactly this.

For information I published this example on elixirforums:

```elixir
def mount(params, session, socket) do
  if not connected?(socket) do
    # Initial request: here we render only **public** data.
    # The response can be cached by shared caches
    products = Products.list(params)

    {:ok, assign(socket, products: products)}
  else
    # Check user authentication
    socket =
      case session do
        %{"user_id" => user_id} ->
          assign(socket, user: Accounts.get_user!(user_id))

        _ ->
          socket
      end

    # Perform stateful stuff that don't depend on the user
    Phoenix.PubSub.subscribe(MyApp.PubSub, "product-updates")

    if socket.assigns[:user] do
      # Prepare rendering for authenticated users
      socket =
        socket
        |> assign(:pending_orders, User.get_pending_orders!(socket.assigns.user))
        |> assign(:pending_notifications, User.get_pending_notifications(socket.assigns.user))

      {:ok, socket}
    else
      # Prepare rendering for anonymous users
      {:ok, socket}
    end
  end
end
```

One of the goal of this PR is to determine if there's a chance to have it in Phoenix, given the security risks it indeed poses.

> I understand that you could use Vary: Cookie as for the HTTP Caching but then I'm not sure what you are meant to gain from HTTP caching with Vary: Cookie on a LiveView.

I think the most common case would be to not set cookie from Liveviews. I just updated `http_cache` yesterday to prevent setting cookies on shared caches.